### PR TITLE
Add tenant policies

### PR DESCRIPTION
**Pull request by ChiaraOggeri**
Co-authored-by: Sofia Munari <sofymunari@gmail.com>

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: policies/verify-tenant-patch/policies/rego_rule.rego [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+violation[{"msg": msg, "details": {}}] {
+	operation := input.review.operation
+	operation == "CREATE"
+	user := input.review.userInfo.username
+	tenantName := input.review.object.metadata.name
+	user == tenantName
+	msg := sprintf("Denied creation request. %v already has a tenant", [user])
+}
+
+#check creation of new user can be done only in workspaces where the creatore is enrolled
```
Comment (giorio94) [inline policies/verify-tenant-patch/policies/rego_rule.rego:20]:
```suggestion
#check creation of new user can be done only in workspaces where the creator is enrolled
```

File: policies/verify-tenant-patch/policies/rule_rego_test.rego [hunk 2]
```diff
@@ -0,0 +1,556 @@
+package crownlabs_tenant_patch
+
+#creator doesn't have any tenant
+test_no_tenant_present {
+	input := {"review": input_review7}
```
Comment (giorio94) [inline policies/verify-tenant-patch/policies/rule_rego_test.rego:5]: Would it be possible to use more explicit names to explain the content of the input_review, to make them easier to understand? e.g. input_review1 -> input_review_creator_has_no_tenant

**Review by giorio94 — COMMENTED** [review]
Thanks! This looks great to me!

I left a couple of minor comments inline. 
In addition, I would ask to
* Try to understand whether it is possible to skip the checks for certain types of identities (e.g. cluster-admin and service accounts)
* Add a couple of test about the deletion of a workspace from a tenant resource, both with success and failure (just to be sure it also works in that case)

File: policies/verify-tenant-patch/policies/rule_rego_test.rego [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+		"object": {
+			"apiVersion": "crownlabs.polito.it/v1alpha1", "kind": "Tenant",
+			"metadata": {"creationTimestamp": "2020-12-25T16:16:14Z", "generation": 1, "name": "test-name", "uid": "c3b3a51a-1a43-47f2-a7f4-c84914b026ba"},
+			"spec": {
+				"email": "test.name@email.com", "firstName": "test-name", "lastName": "test-last", "publicKeys": ["key1", "key2", "key3"],
+				"workspaces": [{"groupNumber": 12, "role": "user", "workspaceRef": {"name": "landa"}}],
+			},
+		},
+		"oldObject": null,
+		"userInfo": {"username": "mariorossi", "groups": ["system:masters", "system:authenticated","kubernetes:admin"]},
```
Comment (giorio94) [inline policies/verify-tenant-patch/policies/rule_rego_test.rego:587]: I would use a different username or workspace in this test, as IIUC, this user can already manage this course according to his Tenant resource.

File: policies/verify-tenant-patch/policies/rule_rego_test.rego [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+	}
+}
+
+
+
+input_review_valid_patch_serviceaccounts = output {
+	output = {
+		"requestKind": {"kind": "Tenant", "group": "crownlabs.polito.it", "version": "v1alpha1"},
+		"operation": "UPDATE",
+		"userInfo": {"username": "mariorossi", "groups": ["system:masters", "system:authenticated","system:serviceaccounts"]},
```
Comment (giorio94) [inline policies/verify-tenant-patch/policies/rule_rego_test.rego:597]: Same as before

File: policies/verify-tenant-patch/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+
```
Comment (giorio94) [inline policies/verify-tenant-patch/README.md:1]: Could you reduce a little the number of newlines? Just to avoid a lengthy file

**Review by giorio94 — APPROVED** [review]
/lgtm

Just a couple of very minor comments inline. Could you also edit the PR description to add a little bit of context?

**Review by palexster — APPROVED** [review]

**Pull request merged**